### ISSUE TITLE
Native GPU version of `tf.dynamic_stitch`

### ISSUE THREAD

**Issue by drasmuss** [tensorflow/tensorflow#7251]
### Environment info
Operating System: Windows 10

Installed version of CUDA and cuDNN: 8.0, 5105
tensorflow release 0.12.1

### If possible, provide a minimal reproducible example (We usually don't have time to read hundreds of lines of your code)
``` python
import tensorflow as tf
from tensorflow.python.client.timeline import Timeline

with tf.device("/gpu:0"):
    x = tf.ones(100)
    idxs = tf.range(100)

    for _ in range(10):
        y = tf.identity(x)
        x = tf.dynamic_stitch([idxs, idxs], [x, y])
        # x = tf.gather(y, idxs)

with tf.Session(config=tf.ConfigProto(log_device_placement=True)) as sess:
    metadata = tf.RunMetadata()
    sess.run(x, options=tf.RunOptions(trace_level=tf.RunOptions.FULL_TRACE),
             run_metadata=metadata)

timeline = Timeline(metadata.step_stats)
with open("profile.json", "w") as f:
    f.write(timeline.generate_chrome_trace_format())
```

The `log_device_placement` output shows that everything is assigned to the GPU, as expected.  However, inspecting the trace output shows that data is being copied on and off the GPU for each call to `dynamic_stitch`.  This is something specific to the `dynamic_stitch` implementation, because using `tf.gather` (a similar indexed read operation, and functionally equivalent in this case), doesn't show this behaviour.

Is this intended behaviour for `dynamic_stitch` (i.e., the copying to and from the GPU is necessary)?  Or is this a bug?  If it isn't a bug, is there some equivalent solution that doesn't require the data to be copied back and forth?

**Comment by yaroslavvb** [comment]
This is probably due to the following `HostMemory` annotations in definition of DynamicStitch in [dynamic_stitch_op.cc](https://github.com/tensorflow/tensorflow/blob/27711108b5fce2e1692f9440631a183b3808fa01/tensorflow/core/kernels/dynamic_stitch_op.cc#L184)

```
#define REGISTER_DYNAMIC_STITCH_GPU(type)                \
  REGISTER_KERNEL_BUILDER(Name("DynamicStitch")          \
                              .Device(DEVICE_GPU)        \
                              .TypeConstraint<type>("T") \
                              .HostMemory("indices")     \
                              .HostMemory("data")        \
                              .HostMemory("merged"),     \
                          DynamicStitchOp<type>)
```

These kinds of annotations are were originally added because the data needed to be on CPU, but you wanted to avoid the extra overhead of scheduling inter-device transfer in TensorFlow. It looks like the code hasn't been touched since open-sourcing, so perhaps they don't make sense anymore, I guess you could try removing them and seeing if it speeds things up?

**Comment by yaroslavvb** [comment]
I looked at it with a bit more instrumentation in [dynamic_stitch_gpu.py](https://github.com/yaroslavvb/stuff/blob/master/dynamic_stitch_gpu.py) and it looks like there are two kinds of transfers going on.

If you run with `output_partition_graphs=True` and look at `partition_graphs` in generated metadata [pbtxt](https://github.com/yaroslavvb/stuff/blob/master/dynamic_stitch_gpu_profile.pbtxt) you see there are ops like this

```
  node {
    name: "x/_0"
    op: "_Send"
    input: "x"
    device: "/job:localhost/replica:0/task:0/gpu:0"
    attr {
      key: "T"
      value {
        type: DT_FLOAT
      }
    }

```

This is the input `x` which is on GPU, and is transferred onto CPU because that input has `HostMemory` attribute, hence needs to be copied.

Later you have ops like this

```
  node {
    name: "stitch-0/_4"
    op: "_HostSend"
    input: "stitch-0"
    device: "/job:localhost/replica:0/task:0/gpu:0"
```

which indicates that output was placed into Host memory, and needs to be transferred to GPU memory. To summarize, it seems like `dynamic_stitch` is not optimized for GPU, and may be faster to run it pinned  to CPU.

**Comment by girving** [comment]
A natively GPU version of `dynamic_stitch` looks doable to me (not fundamentally harder than scatter or gather), so I'll mark this contributions welcome.

**Comment by yaroslavvb** [comment]
One gotcha to avoid is to make sure GPU outputs are real-valued, since integer tensors on GPU don't seem to be supported ([register_types.h](https://github.com/tensorflow/tensorflow/blob/a304537954a865752ad1b18461e6bd67b36082db/tensorflow/core/framework/register_types.h))

```
// Call "m" on all types supported on GPU.
#define TF_CALL_GPU_NUMBER_TYPES(m) \
  TF_CALL_half(m) TF_CALL_float(m) TF_CALL_double(m)

#define TF_CALL_GPU_NUMBER_TYPES_NO_HALF(m) TF_CALL_float(m) TF_CALL_double(m)

```

**Comment by MycChiu** [comment]
I have written a pretty simple Cuda kernel for this op, please see pull request #7764

**Comment by skye** [comment]
@MycChiu did you ever open a new PR for this?

**Comment by nolanliou** [comment]
@girving @skye The [#8260 PR](https://github.com/tensorflow/tensorflow/pull/8260) has been stalled in the past few months, and I also need the GPU version of dynamic_stitch op recently. So may i take over the issue? Thanks.

**Comment by girving** [comment]
@nolanliou Probably, pending @skye's thoughts, but if so it's probably worth describing the algorithm you plan to use.  The previous approach had some fundamental issues.

**Comment by nolanliou** [comment]
@girving
we have to check the collision before sending data to GPU given the  the collision requirements.  
So I wanna divide into two cases:

1. No collision: easy to use threadIdx/blockIdx to index. (most time).
2. Collison: compute all the output's indies at CPU to resolve the collision, then send to GPU.

Welcome any tips.

**Comment by girving** [comment]
@nolanliou It would be better to do it all in one pass, both to avoid the separate collision checking phase and to avoid CPU work in the collision case.  One option is to have a separate buffer storing the source index of an output index.  Atomically writing an entry in the separate buffer at the same time as the main buffer can avoid collisions.  Not sure if the separate buffer can be avoided.

**Comment by Mazecreator** [comment]
This message was created automatically by mail delivery software.

A message that you sent could not be delivered to one or more of its
recipients. This is a temporary error. The following address(es) deferred:

  mazecreator@gmail.com
    Domain mazecreator.com has exceeded the max emails per hour (27/25 (108%)) allowed.  Message will be reattempted later

------- This is a copy of the message, including all the headers. ------
Received: from o8.sgmail.github.com ([167.89.101.199]:55459)
	by server2.lowesthostingrates.com with esmtps (TLSv1.2:ECDHE-RSA-AES128-GCM-SHA256:128)
	(Exim 4.89)
	(envelope-from <bounces+848413-e6d9-mazecreator=mazecreator.com@sgmail.github.com>)
	id 1dcDaF-0004jv-BH
	for mazecreator@mazecreator.com; Mon, 31 Jul 2017 11:30:49 -0500
DKIM-Signature: v=1; a=rsa-sha1; c=relaxed/relaxed; d=github.com; 
	h=from:reply-to:to:cc:in-reply-to:references:subject:mime-version:content-type:content-transfer-encoding:list-id:list-archive:list-post:list-unsubscribe; 
	s=s20150108; bh=nx71yJs08keP5T07fL7PA0zJ5Sk=; b=EWCBHGn0d7W2HWBf
	c31a0Zv8qb4xspIsMQJJX0Ygav/e8OSs9mBlJQAdUPOCa/rPEPAH6hiteaC02key
	8srV34R1lhfmYLYRRNp0hw+i4H/9RQbV4+LfgnBUmnbc5lXkz3tM1w+80TLv7UJ2
	Y7keWQbgc/4OwbfMjTCXQ7cPO1E=
Received: by filter1090p1mdw1.sendgrid.net with SMTP id filter1090p1mdw1-31658-597F5D93-1E
        2017-07-31 16:40:51.938204643 +0000 UTC
Received: from github-smtp2b-ext-cp1-prd.iad.github.net (github-smtp2b-ext-cp1-prd.iad.github.net [192.30.253.17])
	by ismtpd0003p1iad1.sendgrid.net (SG) with ESMTP id ej7d3jnRSfiR-na-eSX06A
	for <mazecreator@mazecreator.com>; Mon, 31 Jul 2017 16:40:51.885 +0000 (UTC)
Date: Mon, 31 Jul 2017 16:40:52 +0000 (UTC)

**Comment by nolanliou** [comment]
@girving Yeah, it could avoid the collision checking, I have implemented the code and post a [PR](https://github.com/tensorflow/tensorflow/pull/11940), please review it if you have time.

**Comment by nolanliou** [comment]
@girving thanks and good luck.

**Comment by yongtang** [comment]
I think this issue has been fixed by PR #11940. Probably could be closed?

**Comment by nolanliou** [comment]
It's ok.
